### PR TITLE
Simplify Plugin API for constructing modules

### PR DESCRIPTION
**Pull request by rjernst**
The Plugin interface currently contains 6 different methods for
adding modules. Elasticsearch has 3 different levels of injectors,
and for each of those, there are two methods. The first takes no
arguments and returns a collection of class objects to construct. The
second takes a Settings object and returns a collection of module
objects already constructed. The settings argument is unecessary because
the plugin can already get the settings from its constructor. Removing
that, the only difference between the two versions is returning an
already constructed Module, or a module Class, and there is no reason
the plugin can't construct all their modules themselves.

This change reduces the plugin api down to just 3 methods for adding
modules. Each returns a Collection<Module>. It also removes the
processModule method, which was unnecessary since onModule
implementations fullfill the same requirement. And finally, it renames
the modules() method to nodeModules() so it is clear these are created
once for each node.

### CODE REVIEW

**Comment by dadoonet** [comment]
I really like the change. Something I'm just wondering. 
Does this PR change anything in the way some components are loaded? 

I mean that before the change plugin components were loaded only when used. For example, `AwsEc2Service` is loaded only if the user set discovery to `ec2`. Is it still the case with this PR? Components are loaded only on demand?

**Comment by es-github-notifications** [comment]
@dadoonet This does not change anything about component loading. Components are still returned as classes and constructed with the injector. This is simply about modules, whose sole purpose is to bind classes the injector will use when necessary.

**Comment by rjernst** [comment]
@dadoonet This does not change anything about component loading. Components are still returned as classes and constructed with the injector. This is simply about modules, whose sole purpose is to bind classes the injector will use when necessary.

**Comment by dadoonet** [comment]
@rjernst Thanks for confirming!
The change looks good to me but I think another reviewer would be great.

**Comment by kimchy** [comment]
I really like this change!. Two notes:

- If we are already changing the `Plugin` interface, I would just merge `AbstractPlugin` into `Plugin` and remove the interface aspect.
- Why did you remove passing Settings to the different modules constructions? Those modules might need the Settings to properly construct? How will they do it now?

**Comment by rjernst** [comment]
@kimchy Plugins have the option of a ctor that takes Settings. They can store this, and pass to any of their modules that need settings.

**Comment by rjernst** [comment]
Ok, settings are needed for the shard/index modules, since those settings are the index settings, not the node settings. I'll add that parameter back.

**Comment by kimchy** [comment]
Yea, it is the shard/index level Settings that are important... . I would still think how to clean up the duplication of which Settings to use, thinking that maybe cleanest is to pass the right Settings to all callbacks (modules/service) and not rely on black magic constructor based Settings injection, which one needs to remember are node based settings. But that is a different change.

**Comment by rjernst** [comment]
@kimchy I merged AbstractPlugin and Plugin, and added back Settings for the index/shard modules methods.

File: core/src/main/java/org/elasticsearch/plugins/AbstractPlugin.java [hunk 1]
```diff
@@ -19,103 +19,17 @@
...
-    @Override
-    public void processModule(Module module) {
-        // nothing to do here
-    }
-
-    @Override
-    public Settings additionalSettings() {
-        return Settings.Builder.EMPTY_SETTINGS;
-    }
+public abstract class AbstractPlugin extends Plugin {
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/plugins/AbstractPlugin.java]: do we really need this class?

**Comment by s1monw** [comment]
added one comment other than that LGTM

File: core/src/main/java/org/elasticsearch/plugins/AbstractPlugin.java (hunk 1 above)
Comment (rjernst) [inline core/src/main/java/org/elasticsearch/plugins/AbstractPlugin.java]: Ooops, meant to remove that.

**Comment by rjernst** [comment]
I will backport to the 2.0 branch once the beta is out.

**Pull request merged**